lxd/storage/backend_lxd.go

```diff
@@ -1344,21 +1350,33 @@ func (b *lxdBackend) UpdateInstanceSnapshot(inst instance.Instance, newDesc stri
 // MigrateInstance sends an instance volume for migration.
 // The args.Name field is ignored and the name of the instance is used instead.
 func (b *lxdBackend) MigrateInstance(inst instance.Instance, conn io.ReadWriteCloser, args *migration.VolumeSourceArgs, op *operations.Operation) error {
+	// This is used for renaming when doing clustering with Ceph.
```

Comment:
@monstermunchkin I'm not clear why this section needs to be here, why can the call to `MigrateVolume` on line 1374 be used for this driver?

Reply:
I moved it to the top as `inst` is nil is some cases which would cause trouble.

Reply:
@monstermunchkin hrm, that doesn't seem right, why is inst nil in some cases, as this function is for migrating instances, instance should always be applied.

Reply:
This is only the case when dealing with clustering in Ceph.

Reply:
@monstermunchkin how strange, can you elaborate on that point, I'm not fully understanding.

The issues with this approach though are that:

1. The position of this exception means we have lost the logging that is just below it in the case where there is an error. For the other drivers, we see the process start via the logging and then if it fails we have some context as to what is happening.
2. The volume type is hard coded to container (but it should pass through the appropriate volume type for the instance, and if the underlying driver doesn't support that instance then it is up to the driver to reject it).
3. The volume's name is derived from the instance's project and the args.Name, this ensures that drivers support projects (there were quite a few project related bugs in the old storage layer). The `Volume` type should always have a volume name that is that reflected on the storage device itself (i.e it should contain the project name for instance volumes). So in this case the `Volume` being created via `newVolume()` is just passing args.Name and so is not project compatible.
4. The root disk config from the instance is not being passed through, which is the case for all other driver types.

Reply:
Ceph supports moving a container within the cluster even if one of the nodes is offline. In this case, `inst` will be `nil`. However, this is not a problem.

> 1. The position of this exception means we have lost the logging that is just below it in the case where there is an error. For the other drivers, we see the process start via the logging and then if it fails we have some context as to what is happening.

We could adjust the `log.Ctx` depending on `inst`, then we could have logging even for this special case.

> 2. The volume type is hard coded to container (but it should pass through the appropriate volume type for the instance, and if the underlying driver doesn't support that volume type then it is up to the driver to reject it).

Since `inst` might be `nil`, the volume type would need to be passed differently.

> 3. The Volume's name is derived from the instance's project and the args.Name, this ensures that drivers support projects (there were quite a few project related bugs in the old storage layer). The Volume type should always have a volume name that is that reflected on the storage device itself (i.e it should contain the project name for instance volumes). So in this case the Volume being created via newVolume() is just passing args.Name and so is not project compatible.

True. This can be fixed by using `project.Prefix(project, oldName)` and `project.Prefix(project, newName)` in the `VolumeSourceArgs`.

> 4. The root disk config from the instance is not being passed through, which is the case for all other driver types.

This disk config is not of interest. Ceph will only perform an `unmap` and/or `rename` action which only requires pool and volume names.

Reply:
@monstermunchkin can you point me in the direction of the code where this function is called with a nil instance, I'm wondering if we should instead create a 'fake' instance (of the correct type), and pass it in that case, thus moving the special case to the ceph code rather than have to deal with it in generic backendLXD code.

When one of the nodes is offline, presumably the DB is still accessible from the other nodes, so DB lookups should work right?

Reply:
Loading the instance: https://github.com/lxc/lxd/blob/master/lxd/container_post.go#L131-L145
Calling `containerPostClusteringMigrateWithCeph`: https://github.com/lxc/lxd/blob/master/lxd/container_post.go#L193

Yes, the DB works in this case.

Reply:
So from what I can see in the `containerPost` function, you should be able to remove the `var inst instance.Instance` and move this party `inst, err = instance.LoadByProjectAndName(d.State(), project, name)` up to line 146.

At that point this logic breaks, as inst will never be nil (what we want) `if inst != nil && inst.IsRunning()` but presumably this was added because we don't want to check if a container is running if we know the source node is offline. This can probably be modified to only check if running if the source is online rather than check if inst is nil.

There is a comment above that says 'we don't want to load the container here (since  it's not a local container)' but I'm not sure why (even if we dont attempt to start the container), is that not allowed @stgraber ?

But assuming we can load the container in all circumstances, then you should be OK to pass around a proper instantiated instance.

Reply:
It should be fine to load, operations on it will be broken but just loading the struct should be okay.

Reply:
(It wasn't always the case, we used to generate the full LXC config and access some on-disk stuff just during load a while back)

Reply:
Fixed